aws/macie.go

```diff
@@ -0,0 +1,87 @@
+package aws
+
+import (
+	goerror "errors"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/macie2"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/go-commons/errors"
+)
+
+func getAllMacieAccounts(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]string, error) {
+	svc := macie2.New(session)
+	stssvc := sts.New(session)
+
+	allMacieAccounts := []string{}
+	output, err := svc.GetAdministratorAccount(&macie2.GetAdministratorAccountInput{})
+	// If the current account does have an Administrator account relationship, and it is enabled, then we consider this a macie member account
+	if output.Administrator != nil && output.Administrator.RelationshipStatus != nil {
+		if aws.StringValue(output.Administrator.RelationshipStatus) == macie2.RelationshipStatusEnabled {
+
+			input := &sts.GetCallerIdentityInput{}
+			output, err := stssvc.GetCallerIdentity(input)
+			if err != nil {
+				return allMacieAccounts, errors.WithStackTrace(err)
+			}
+
+			currentAccountId := aws.StringValue(output.Account)
+
+			allMacieAccounts = append(allMacieAccounts, currentAccountId)
+		}
+	}
+	if err != nil {
+
+		// There are several different errors that AWS may return when you attempt to call Macie operations on an account
+		// that doesn't yet have Macie enabled. For our purposes, this is fine, as we're only looking for those accounts and
+		// regions where Macie is enabled. Therefore, we ignore only these expected errors, and return any other errror that might occur
+		var ade *macie2.AccessDeniedException
+		var rnfe *macie2.ResourceNotFoundException
+
+		switch {
+		case goerror.As(err, &ade):
+			logging.Logger.Debugf("Macie AccessDeniedException means macie is not enabled in account, so skipping")
+		case goerror.As(err, &rnfe):
+			logging.Logger.Debugf("Macie ResourceNotFoundException means macie is not enabled in account, so skipping")
+		default:
+			return allMacieAccounts, errors.WithStackTrace(err)
+		}
```

Comment:
This condition check should happen first, and should return `nil, nil` at the end. The reason is because we don't want to touch `output` at all if there is an error since the structure of the object is undefined, so to prevent any potential bugs, we should always check the error first.